futures-util/src/future/shared.rs

```diff
@@ -153,17 +158,35 @@ impl<Fut> Shared<Fut> where Fut: Future {
 
     /// Safety: callers must first ensure that `self.inner.state`
     /// is `COMPLETE`
-    unsafe fn clone_output(&self) -> Arc<Fut::Output> {
-        if let FutureOrOutput::Output(item) = &*self.inner.future_or_output.get() {
-            item.clone()
-        } else {
-            unreachable!()
+    unsafe fn try_take_output(&mut self) -> Fut::Output {
```

Comment:
`take_or_clone_output()` would be better. 

(The convention is that functions prefixed with `try_` should return something the implements the `Try` trait - often `Result`)